crates/store/re_query2/src/latest_at.rs

```diff
@@ -0,0 +1,681 @@
+///
+/// Use [`LatestAtResults::get`] and/or [`LatestAtResults::get_required`] in order to access
+/// the results for each individual component.
+#[derive(Debug)]
+pub struct LatestAtResults {
+    /// The associated [`EntityPath`].
+    pub entity_path: EntityPath,
+
+    /// The query that yielded these results.
+    pub query: LatestAtQuery,
+
+    /// The compound index of this query result.
+    ///
+    /// A latest-at query is a compound operation that gathers data from many different rows.
+    /// The index of that compound result corresponds to the index of most the recent row in all the
+    /// sub-results, as defined by time and row-id order.
+    pub compound_index: (TimeInt, RowId),
+
+    /// Results for each individual component.
+    pub components: IntMap<ComponentName, UnitChunkShared>,
+}
+
+impl LatestAtResults {
+    #[inline]
+    pub fn empty(entity_path: EntityPath, query: LatestAtQuery) -> Self {
+        Self {
+            entity_path,
+            query,
+            compound_index: (TimeInt::STATIC, RowId::ZERO),
+            components: Default::default(),
+        }
+    }
+}
+
+impl LatestAtResults {
+    #[inline]
+    pub fn contains(&self, component_name: &ComponentName) -> bool {
+        self.components.contains_key(component_name)
+    }
+
+    /// Returns the [`UnitChunkShared`] for the specified [`Component`].
+    #[inline]
+    pub fn get(&self, component_name: &ComponentName) -> Option<&UnitChunkShared> {
+        self.components.get(component_name)
+    }
+
+    /// Returns the [`UnitChunkShared`] for the specified [`Component`].
+    ///
+    /// Returns an error if the component is not present.
+    #[inline]
+    pub fn get_required(&self, component_name: &ComponentName) -> crate::Result<&UnitChunkShared> {
+        if let Some(component) = self.get(component_name) {
+            Ok(component)
+        } else {
+            Err(QueryError::PrimaryNotFound(*component_name))
+        }
+    }
+
+    /// Returns the compound index (`(TimeInt, RowId)` pair) of the results.
+    #[inline]
+    pub fn index(&self) -> (TimeInt, RowId) {
+        self.compound_index
+    }
+}
+
+impl LatestAtResults {
+    #[doc(hidden)]
+    #[inline]
+    pub fn add(
+        &mut self,
+        component_name: ComponentName,
+        index: (TimeInt, RowId),
+        chunk: UnitChunkShared,
+    ) {
+        debug_assert!(chunk.num_rows() == 1);
+
+        // NOTE: Since this is a compound API that actually emits multiple queries, the index of the
+        // final result is the most recent index among all of its components, as defined by time
+        // and row-id order.
+        if index > self.compound_index {
+            self.compound_index = index;
+        }
+
+        self.components.insert(component_name, chunk);
+    }
+}
+
+// --- Helpers ---
+//
+// Helpers for UI and other high-level/user-facing code.
+//
+// In particular, these replace all error handling with logs instead.
+
+impl LatestAtResults {
```

Comment:
:point_down: All the nice component-level helpers that mirror those on `UnitChunk`, but specifically thought with UI/viewer code in mind.